### PR TITLE
fix: save per-run MSA outputs

### PR DESCRIPTION
**Pull request by ahmedtaha100**
## Summary

This updates the MSA output workflow discussed in [aqlaboratory/openfold-3#31](https://github.com/aqlaboratory/openfold-3/issues/31).

MSA preparation now uses a unique temporary workspace. During prediction, OpenFold saves both the raw ColabFold files and its formatted alignments by default, with separate settings controlling which records prediction keeps. The temporary workspace is always removed during final cleanup.

## Changes

- Save raw ColabFold files as soon as the server response has been validated, before OpenFold starts formatting it.
- Save OpenFold-formatted alignments after MSA processing finishes.
- Add `save_openfold_outputs` and `save_colabfold_outputs` settings, both enabled by default.
- Add `colabfold_output_dir` for choosing a different parent directory for raw records.
- `cleanup_msa_dir` now controls temporary template preprocessing cleanup. The new save settings control which MSA records are kept.
- Keep existing explicit `msa_output_directory` destinations separate from the temporary workspace and reject paths that cleanup could delete.
- Keep `align-msa-server` outputs and the paths written to `query_msa.json` after temporary cleanup.
- Update the configuration reference, inference guide, and examples.

## Related Issues

Related to [aqlaboratory/openfold-3#31](https://github.com/aqlaboratory/openfold-3/issues/31).

## Testing

- Focused affected suite: `96 passed, 4 deselected`. The four deselected tests require external S3 setup and are unrelated to this change.
- Ruff formatting and lint passed.
- Mypy passed against the current baseline.
- All three changed YAML examples parsed successfully.
- Strict Sphinx documentation build passed.
- `git diff --check` passed.
- CUDA 13 validation on an NVIDIA GeForce RTX 5090 passed for server-backed prediction, prediction without the MSA server, and `align-msa-server`. The server-backed and alignment-only runs kept their saved outputs and removed their temporary workspaces. The no-server run saved the dummy formatted MSA without creating raw output.

### CODE REVIEW

File: openfold3/core/data/tools/colabfold_msa_server.py [hunk 1]
```diff
@@ -1200,20 +1205,22 @@ def preprocess_colabfold_msas(
     output_directory = compute_settings.msa_output_directory
     logger.warning(f"Using output directory: {output_directory} for ColabFold MSAs.")
 
+    raw_dir = output_directory / "raw"
+    try:
+        # Creating `raw/` is the reservation. A second process fails here.
+        raw_dir.mkdir(parents=True, exist_ok=False)
+    except FileExistsError:
+        raise FileExistsError(
```
Comment (jnwei) [inline openfold3/core/data/tools/colabfold_msa_server.py]: Moving the `raw` directory test here, before the mappings are generated is a good move.

File: openfold3/tests/core/data/tools/test_colabfold_msa_server.py
Comment (jnwei) [inline openfold3/tests/core/data/tools/test_colabfold_msa_server.py:1]: Good test set up with appropriate mocks / dummy values

**Review by jnwei — CHANGES_REQUESTED** [review]
@ahmedtaha100 Thank you for this excellent contribution. I appreciate that you have added documentation and also extensive unit tests.

I will leave some minor comments here, but I'd like to have a deeper discussion about what the best approach is for handling the colab raw directory on the parent issue: https://github.com/aqlaboratory/openfold-3/issues/31. I'm afraid that the current approach in this PR does not work well for users with large numbers of queries that do not use colabfold outputs.

Let me know if you agree with the alternative approach outlined in https://github.com/aqlaboratory/openfold-3/issues/31, or if you have any suggestions on how to improve the approach.

**Comment by ahmedtaha100** [comment]
> @ahmedtaha100 Thank you for this excellent contribution. I appreciate that you have added documentation and also extensive unit tests.
> 
> I will leave some minor comments here, but I'd like to have a deeper discussion about what the best approach is for handling the colab raw directory on the parent issue: #31. I'm afraid that the current approach in this PR does not work well for users with large numbers of queries that do not use colabfold outputs.
> 
> Let me know if you agree with the alternative approach outlined in #31, or if you have any suggestions on how to improve the approach.

Thank you for your feedback, @jnwei. I agree with the approach you proposed in #31.

The current PR is awkward for people who run repeated jobs against the same MSA directory and only want OpenFold's organized alignments. I agree that `colabfold_output_dir` should control whether the raw output is copied elsewhere, and that the working `raw/` directory should be removed whether or not a copy is requested.

I left a fuller reply on #31. The two points I raised there are:

1. The default behavior. Your proposal makes raw preservation opt-in by defaulting `colabfold_output_dir` to `None`. @jandom suggested preserving everything by default while keeping it configurable for large-scale workloads, which is the option I prefer.
2. When to copy the raw output. Your proposal waits until the full inference job succeeds. I think it is worth copying the raw output once MSA processing succeeds so a later model failure does not prevent it from being preserved.

I'll keep the early `raw/` reservation and the tests around it. Which default and copy point would you prefer? Once we figure out those two details, I'll update the PR accordingly.

File: openfold3/core/data/tools/colabfold_msa_server.py [hunk 2]
```diff
@@ -1088,17 +1123,113 @@ class MsaComputationSettings(BaseModel):
...
-    def create_dir(self) -> "MsaComputationSettings":
-        """Creates the output directory if it does not exist."""
-        if self.msa_output_directory is None:
-            from openfold3.core.data.tools.utils import get_of3_tmpdir
+    def create_workspace(self) -> None:
+        """Create and mark the temporary workspace owned by this run."""
+        self.workspace_directory.mkdir(parents=True, exist_ok=False)
+        try:
+            (self.workspace_directory / ".openfold3-msa-workspace").write_text(
+                self._workspace_token
```
Comment (jnwei) [inline openfold3/core/data/tools/colabfold_msa_server.py]: Is the purpose of the `_workspace_token` primarily to ensure that the user is able to write and delete the specified `workspace_directory`? Would it work to write "test" or the current date/time?

I was unfamiliar with this pattern so I did some research. It looks like this pattern is typically used in security practices to prevent (malicious) third parties from altering the input directories, e.g. TOCTOU attacks. I'm not sure if we need to follow the same practice here, but perhaps I am missing a use case?

File: openfold3/core/data/tools/colabfold_msa_server.py [hunk 3]
```diff
@@ -1088,17 +1123,113 @@ class MsaComputationSettings(BaseModel):
     save_mappings: bool = True
     msa_output_directory: Path | None = None
     cleanup_msa_dir: bool = True
+    save_openfold_outputs: bool = True
+    save_colabfold_outputs: bool = True
+    colabfold_output_dir: Path | None = None
+    _workspace_token: str = PrivateAttr(default_factory=lambda: uuid4().hex)
+    _run_directory_name: str = PrivateAttr(default_factory=lambda: f"msa-{uuid4().hex}")
```
Comment (jnwei) [inline openfold3/core/data/tools/colabfold_msa_server.py]: if possible, it would be good to include the user name in the default _run_directory_name, and possibly the date-time instead of the uuid hex

File: openfold3/core/data/tools/colabfold_msa_server.py [hunk 4]
```diff
@@ -1088,17 +1123,113 @@ class MsaComputationSettings(BaseModel):
...
+                removed = not self.workspace_directory.exists()
+                if not removed:
+                    logger.warning(
+                        f"Could not remove temporary MSA directory: "
+                        f"{self.workspace_directory}"
+                    )
+                return removed
+        except OSError:
+            pass
+        return False
```
Comment (jnwei) [inline openfold3/core/data/tools/colabfold_msa_server.py]: This check of without the _workspace_token will be simpler as it would just check for the existance fo the directory.

Could we change this logic to so that rather than returning a bool, it raises the error directly?

Then the block should be wrapped in a finally-try-except block, e.g.

```
finally:
    try:
        shutil.rmtree(self.workspace_directory)
    except OSError:
        logger.warning(f"Could not remove temporary MSA directory: ...", exc_info=True)
```

File: openfold3/tests/core/data/tools/test_colabfold_msa_server.py [hunk 5]
```diff
@@ -321,6 +334,106 @@ def test_runner_on_multimer_example(
...
+        assert saved_raw.exists() is save_colabfold
+        assert (saved_output / "mappings/seq_to_rep_id.json").exists() is save_openfold
+        expected_root = saved_output if save_openfold else workspace
+        assert (
+            processed.queries["query1"]
+            .chains[0]
+            .main_msa_file_paths[0]
+            .is_relative_to(expected_root)
+        )
+        assert settings.cleanup_workspace()
```
Comment (jnwei) [inline openfold3/tests/core/data/tools/test_colabfold_msa_server.py]: This was called on its own in the previous test, and with the assert here. I think it's preferrable to call without the assert, especially if cleanup_workspace() is changed to provide errors rather than a bool

File: openfold3/tests/core/data/tools/test_colabfold_msa_server.py [hunk 6]
```diff
@@ -321,6 +334,106 @@ def test_runner_on_multimer_example(
...
+        query = self._construct_monomer_query("TEST")
+        processed = preprocess_colabfold_msas(query, settings)
+        saved_main = (
+            saved_output / "main" / get_sequence_hash("TEST") / "colabfold_main.a3m"
+        )
+        saved_raw = settings.saved_colabfold_output_directory / "main/server-output.a3m"
+
+        assert saved_main.exists() is save_openfold
+        assert saved_raw.exists() is save_colabfold
+        assert (saved_output / "mappings/seq_to_rep_id.json").exists() is save_openfold
```
Comment (jnwei) [inline openfold3/tests/core/data/tools/test_colabfold_msa_server.py:403]: The check of the `saved_main` and the `saved_raw` should happen after the cleanup step.

File: openfold3/tests/core/data/tools/test_colabfold_msa_server.py [hunk 7]
```diff
@@ -707,27 +845,62 @@ def test_cli_output_dir_overrides_config(self, tmp_path):
...
-        with pytest.raises(ValueError) as exc_info:
+        with pytest.raises(ValueError, match="Output directory mismatch"):
             MsaComputationSettings.from_config_with_cli_override(
-                cli_output_dir, test_yaml_file
+                tmp_path / "cli-output", config_file
             )
 
-        assert "Output directory mismatch" in str(exc_info.value), (
-            "Expected ValueError on output directory conflict"
+    def test_workspace_and_default_records_share_one_run_id(self, tmp_path):
```
Comment (jnwei) [inline openfold3/tests/core/data/tools/test_colabfold_msa_server.py]: nit: this is less about sharing one_run_id and more about sharing the saved_output_root? Perhaps the title of the test could reflect this?

It might also be good to test this at a higher level, e.g. when the `MsaComputationSettings.msa_output_directory` is specified, rather than using the lower level `set_saved_output_root` which the user will typically not use directly

Also, is there a reason to use the InferenceExperimentConfig instead of the MsaComputationSettings directly?

File: openfold3/tests/core/data/tools/test_colabfold_msa_server.py [hunk 8]
```diff
@@ -707,27 +845,62 @@ def test_cli_output_dir_overrides_config(self, tmp_path):
...
+        assert settings.workspace_directory.name == settings.run_directory_name
+        assert settings.saved_output_directory.name == settings.run_directory_name
+        assert (
+            settings.saved_colabfold_output_directory.name
+            == settings.run_directory_name
         )
+        assert other_settings.run_directory_name != settings.run_directory_name
+
+    @pytest.mark.parametrize("nested", [False, True])
+    def test_workspace_cannot_contain_openfold_output(self, nested):
```
Comment (jnwei) [inline openfold3/tests/core/data/tools/test_colabfold_msa_server.py:907]: Nice corner case to test for!

**Review by jnwei — CHANGES_REQUESTED** [review]
Thank you @ahmedtaha100 for taking on the daunting task of reworking the MSA saving behavior! The changes involved are quite complex and touch many different parts of the codebase, some of which had missing test coverage. 

I really appreciate your attention to detail with the test cases and covering many corner cases to ensure that that the test directories cannot be accidentally deleted. Great work!

I have a few mostly small comments, but overall, this PR is in good shape!

File: openfold3/core/data/tools/colabfold_msa_server.py (hunk 2 above)
Comment (ahmedtaha100) [reply openfold3/core/data/tools/colabfold_msa_server.py]: You're right, we don't need the token for a security or TOCTOU use case here. I was only using it as an ownership marker so cleanup would not delete a directory this run did not create. I've removed it and now track successful workspace creation in memory instead.

File: openfold3/core/data/tools/colabfold_msa_server.py (hunk 3 above)
Comment (ahmedtaha100) [reply openfold3/core/data/tools/colabfold_msa_server.py]: Good suggestion. I updated the default to `msa-<user>-<UTC timestamp>-<short suffix>`. The suffix prevents a collision if two runs start at the same time.

File: openfold3/core/data/tools/colabfold_msa_server.py (hunk 4 above)
Comment (ahmedtaha100) [reply openfold3/core/data/tools/colabfold_msa_server.py]: Done. `cleanup_workspace()` now returns nothing and lets deletion errors propagate. Both cleanup paths catch `OSError` and log with `exc_info=True`. I kept a small in-memory flag so a failed workspace creation cannot cause cleanup to delete a directory this run did not create.

File: openfold3/tests/core/data/tools/test_colabfold_msa_server.py (hunk 5 above)
Comment (ahmedtaha100) [reply openfold3/tests/core/data/tools/test_colabfold_msa_server.py]: Agreed. I removed the assertion and now call `cleanup_workspace()` directly, since failures are raised instead of returned as a Boolean.

File: openfold3/tests/core/data/tools/test_colabfold_msa_server.py (hunk 6 above)
Comment (ahmedtaha100) [reply openfold3/tests/core/data/tools/test_colabfold_msa_server.py:403]: Done

File: openfold3/tests/core/data/tools/test_colabfold_msa_server.py (hunk 7 above)
Comment (ahmedtaha100) [reply openfold3/tests/core/data/tools/test_colabfold_msa_server.py]: Agreed. I renamed this test and rewrote it to use `MsaComputationSettings` directly with `msa_output_directory`. This test no longer uses `InferenceExperimentConfig` or calls `set_saved_output_root()`.

File: openfold3/tests/core/data/tools/test_colabfold_msa_server.py [hunk 9]
```diff
@@ -693,12 +813,37 @@ def test_obsolete_pdb_falls_back(self, mock_fetch):
...
+    def test_workspace_cleanup_is_unconditional(self, cleanup_msa_dir):
+        settings = MsaComputationSettings(cleanup_msa_dir=cleanup_msa_dir)
+        settings.create_workspace()
+
+        settings.cleanup_workspace()
+
+        assert not settings.workspace_directory.exists()
+        settings.cleanup_workspace()
+        assert not settings.workspace_directory.exists()
+
```
Comment (jnwei) [inline openfold3/tests/core/data/tools/test_colabfold_msa_server.py]: Is this duplication intentional?

**Review by jnwei — COMMENTED** [review]
This looks great, LGTM! Thanks @ahmedtaha :)

Two very small changes:
- Could you help remove the old cleanup function in ColabFoldQueryRunner, ie. https://github.com/ahmedtaha100/openfold-3/blob/44e6a866b12c3c9fcdd9dbdcd1f94df0da12cb07/openfold3/core/data/tools/colabfold_msa_server.py#L994-L1001 This is unused and redundant with the workspace handling you added to the MSAComputationSettings, and so we should remove it to avoid confusion.
- I had a small question regardina. duplicated cleanup /assert statement in the tests, see comments

**Review by jnwei — APPROVED** [review]

**Review by jnwei — CHANGES_REQUESTED** [review]
Ah sorry, I forgot one more thing: We need to update the inference integration tests with the expected paths now that we've changed the defaults to include the msa directories

https://github.com/ahmedtaha100/openfold-3/blob/fix/issue-31-preserve-colabfold-raw/openfold3/tests/inference/test_inference_full.py#L545

It should be a matter of adding the expected directories for the alignment files to the list of expected directories to check. Let me know if you'd like a hand with this.

**Review by jnwei — APPROVED** [review]
LGTM thank you @ahmedtaha100 This was a prickly issue to resolve, thank you for your hard work on the fix!
